libdnf/base/goal.cpp

```diff
@@ -197,291 +221,373 @@ void Goal::add_module_enable(const std::string & spec) {
+}
+
+void Goal::Impl::add_rpm_ids(
+    libdnf::Goal::Action action, const libdnf::rpm::Package & rpm_package, const libdnf::GoalJobSettings & settings) {
+    if (rpm_package.sack.get() != &base->get_rpm_solv_sack()) {
         throw UsedDifferentSack();
     }
     libdnf::rpm::solv::IdQueue ids;
     ids.push_back(rpm_package.get_id().id);
-    p_impl->rpm_ids.push_back(std::make_tuple(Action::UPGRADE, std::move(ids), false));
+    rpm_ids.push_back(std::make_tuple(action, std::move(ids), settings));
 }
 
-void Goal::add_rpm_upgrade(const libdnf::rpm::PackageSet & package_set) {
-    if (package_set.get_sack() != &p_impl->base->get_rpm_solv_sack()) {
+void Goal::Impl::add_rpm_ids(
+    libdnf::Goal::Action action,
+    const libdnf::rpm::PackageSet & package_set,
+    const libdnf::GoalJobSettings & settings) {
+    if (package_set.get_sack() != &base->get_rpm_solv_sack()) {
         throw UsedDifferentSack();
     }
     libdnf::rpm::solv::IdQueue ids;
     for (auto package_id : *package_set.p_impl) {
         ids.push_back(package_id.id);
     }
-    p_impl->rpm_ids.push_back(std::make_tuple(Action::UPGRADE, std::move(ids), false));
+    rpm_ids.push_back(std::make_tuple(action, std::move(ids), settings));
 }
 
-void Goal::Impl::add_install_to_goal() {
+GoalProblem Goal::Impl::add_specs_to_goal() {
+    auto & sack = base->get_rpm_solv_sack();
+    auto & cfg_main = base->get_config();
+    auto ret = GoalProblem::NO_PROBLEM;
+    for (auto & [action, spec, settings] : rpm_specs) {
+        switch (action) {
+            case Action::INSTALL:
+                ret |= add_install_to_goal(spec, settings);
+                break;
+            case Action::REMOVE:
+                add_remove_to_goal(spec, settings);
+                break;
+            case Action::DISTRO_SYNC:
+            case Action::UPGRADE:
+                add_upgrades_distrosync_to_goal(action, spec, settings);
+                break;
+            case Action::UPGRADE_ALL: {
+                libdnf::rpm::SolvQuery query(&sack);
+                libdnf::rpm::solv::IdQueue upgrade_ids;
+                for (auto package_id : *query.p_impl) {
+                    upgrade_ids.push_back(package_id.id);
+                }
+                bool clean_requirements_on_remove =
+                    settings.clean_requirements_on_remove == GoalSetting::SET_TRUE ? true : false;
+                rpm_goal.add_upgrade(upgrade_ids, settings.get_best(cfg_main), clean_requirements_on_remove);
+            } break;
+            case Action::DISTRO_SYNC_ALL: {
+                libdnf::rpm::SolvQuery query(&sack);
+                libdnf::rpm::solv::IdQueue upgrade_ids;
+                for (auto package_id : *query.p_impl) {
+                    upgrade_ids.push_back(package_id.id);
+                }
+                bool clean_requirements_on_remove =
+                    settings.clean_requirements_on_remove == GoalSetting::SET_TRUE ? true : false;
```

Comment:
Condition `settings.clean_requirements_on_remove == GoalSetting::SET_TRUE` returns `true` or `false`.
So, `bool clean_requirements_on_remove = settings.clean_requirements_on_remove == GoalSetting::SET_TRUE;` is enought.